File: 14주차/fine/HouseRobber.swift
func rob(_ nums: [Int]) -> Int {
    let n = nums.count
    
    if n == 0 {
        return 0
    }

    if n == 1 {
        return nums[0]
    }
    
    var prev2 = nums[0]  // dp[i-2]
    var prev1 = max(nums[0], nums[1])  // dp[i-1]
    
    for i in 2..<n {
        let current = max(prev1, nums[i] + prev2)
        prev2 = prev1  // dp[i-2] 갱신
        prev1 = current  // dp[i-1] 갱신
    }
    
    return prev1  
}

File: 14주차/Evie/House Robber.js
// DP Bottom-Up
// 작은 값들을 누적시키면서 진행

class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    rob(nums) {
        if (nums.length === 0) return 0;
        if (nums.length === 1) return nums[0];

        let prev1 = 0, prev2 = 0;

        for (let num of nums) {
            let temp = prev1;
            prev1 = Math.max(prev1, prev2 + num);
            prev2 = temp;
        }

        return prev1;
    }
}

/*
Swift

class Solution {
    func rob(_ nums: [Int]) -> Int {
        if nums.isEmpty { return 0 }
        if nums.count == 1 { return nums[0] }

        var prev1 = 0
        var prev2 = 0

        for num in nums {
            let temp = prev1
            prev1 = max(prev1, prev2 + num)
            prev2 = temp
        }

        return prev1
    }
}
*/